File: 295-find-median-from-data-stream/295-find-median-from-data-stream.cpp
class MedianFinder {
    priority_queue<int> maxHeap;
    priority_queue<int, vector<int>, greater<int>> minHeap;
public:
    MedianFinder() {
    }
    
    void addNum(int num) {
        if(!minHeap.empty() && num < minHeap.top())
            maxHeap.push(num);
        else
            minHeap.push(num);
        
        int diff = minHeap.size() - maxHeap.size();
        
        if(diff > 1) {
            maxHeap.push(minHeap.top());
            minHeap.pop();
        }
        else if(diff < -1) {
            minHeap.push(maxHeap.top());
            maxHeap.pop();
        }
        
    }
    
    double findMedian() {
        
        if(maxHeap.size() > minHeap.size())
            return (double)maxHeap.top();
        
        else if(maxHeap.size() < minHeap.size())
            return (double)minHeap.top();
        
        else
            return (minHeap.top() + maxHeap.top()) / 2.0;
            
    }
};

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder* obj = new MedianFinder();
 * obj->addNum(num);
 * double param_2 = obj->findMedian();
 */

File: 295-find-median-from-data-stream/295-find-median-from-data-stream.py
class MedianFinder:

    def __init__(self):
        self.minHeap = []
        self.maxHeap = []

    def addNum(self, num: int) -> None:
        if len(self.minHeap) > 0 and num < self.minHeap[0]:
            heappush(self.maxHeap, -num)
        
        else:
            heappush(self.minHeap, num)
        
        diff = len(self.minHeap) - len(self.maxHeap)
        
        if diff > 1:
            heappush(self.maxHeap, -heappop(self.minHeap))
        elif diff < -1:
            heappush(self.minHeap, -heappop(self.maxHeap))

    def findMedian(self) -> float:
        if len(self.minHeap) > len(self.maxHeap):
            return self.minHeap[0]
        
        elif len(self.maxHeap) > len(self.minHeap):
            return -self.maxHeap[0]
    
        else:
            return (self.minHeap[0] - self.maxHeap[0]) / 2


# Your MedianFinder object will be instantiated and called as such:
# obj = MedianFinder()
# obj.addNum(num)
# param_2 = obj.findMedian()